implement `StreamMuxer` for `Connection`; implement `Substream` type

noot opened the pull request:
## Changes

<!--

Please provide a brief but specific list of changes made, describe the changes
in functionality rather than the changes in code.

-->

- implement `StreamMuxer` for `Connection`
- implement `Substream` which implements `AsyncRead + AsyncWrite`; allows for reading, writing, and closing of the stream
- move testcontainers test setup into a macro to reduce copypaste in every test

## Tests

<!--

Details on how to run tests relevant to the changes within this pull request.

-->

```
cargo test
```

## Issues

<!--

Please link any issues that this pull request is related to and use the GitHub
supported format for automatically closing issues (ie, closes #123, fixes #123)

-->

- closes #6

ivanceras reviewed (approved):
lgtm!

ec2 commented on src/connection.rs:
```diff
@@ -50,28 +106,211 @@ impl Connection {
+        self.new_stream_with_id(substream_id)
+    }
+
+    pub fn new_stream_with_id(
+        &mut self,
+        substream_id: SubstreamId,
+    ) -> Result<BoxFuture<'static, Result<(), Error>>, Error> {
```
Same on line 269 in Transport. You are using a 'static lifetime there, but you can just do
`type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;`

ec2 commented:
In general, I think you should try (if you can) to remove some more of your code thats meant for testing out of the actual implementation.

noot commented on src/mixnet.rs:
```diff
@@ -169,32 +175,28 @@ fn parse_nym_message(msg: Message) -> Result<ServerResponse, Error> {
+    use crate::new_nym_client;
 
     #[tokio::test]
     async fn test_mixnet_poll_inbound_and_outbound() {
```
this seems to be a testcontainers issue :/ @stonecharioteer any ideas?

noot commented on src/transport.rs:
```diff
@@ -62,7 +63,24 @@ pub struct NymTransport {
+
+    async fn new_maybe_with_notify_inbound(
+        uri: &String,
+        notify_inbound_tx: Option<UnboundedSender<()>>,
```
no it's used in `NymTransport::new()` also

noot commented on src/message.rs:
```diff
@@ -25,6 +26,25 @@ impl ConnectionId {
+pub struct SubstreamId(pub(crate) [u8; 32]);
+
+impl SubstreamId {
+    pub(crate) fn generate() -> Self {
```
it's not just used for testing, it's used in `new_stream` which I'm pretty sure libp2p will need

noot commented on src/connection.rs:
```diff
@@ -50,28 +106,211 @@ impl Connection {
     }
+
+    /// attempts to open a new stream over the connection; returns a future that resolves when the stream is established.
+    pub fn new_stream(&mut self) -> Result<BoxFuture<'static, Result<(), Error>>, Error> {
```
yeah it's not right now but I think libp2p needs this, since there's no way to open a stream otherwise.. however the trait doesn't explicitly say :|

noot commented:
@ec2 nice, comments addressed, moved the test-only functions inside the test module. also remove the cursed `InnerConnection` type since it wasn't needed anymore, replaced it with one channel :D
